### ISSUE TITLE
Problem with myGDP.getAttributes(OKshapeFile)

### ISSUE THREAD

**Issue by rsignell-usgs** [USGS-CIDA/pyGDP#120]
Updated to 1.3.3, but my Dust Bowl test still failing.

cell [10] error:
http://nbviewer.ipython.org/gist/rsignell-usgs/6b887eaf194e312aa033

**Comment by dblodgett-usgs** [comment]
I'm not sure when all these issues were introduced... don't have time to debug right now, sorry.

**Comment by rsignell-usgs** [comment]
Dave,
No problem.  I'll just tell the USGS Water guys on Friday, "hey, this CIDA
stuff used to work.  Honest it did!"   ;-)

-Rich


On Wed, Sep 2, 2015 at 1:50 PM, David Blodgett <notifications@github.com>
wrote:

> I'm not sure when all these issues were introduced... don't have time to
> debug right now, sorry.
>
> —
> Reply to this email directly or view it on GitHub.
>



-- 
Dr. Richard P. Signell   (508) 457-2229
USGS, 384 Woods Hole Rd.
Woods Hole, MA 02543-1598

**Comment by jread-usgs** [comment]
:cactus:

**Comment by dblodgett-usgs** [comment]
There is actually a test that you can see passing. 

It appears that however you are installing owslib, lxml isn't getting included. If you'd like we can help trouble shoot your install method.

**Comment by rsignell-usgs** [comment]
@dblodgett-usgs , I'm glad this is an environment problem.   Except that I'll have to retract my snide remarks?  ;-)

I'm not sure what you mean, however.  My conda environment definitely has `lxml` as you can see in cell [3] here:
http://nbviewer.ipython.org/gist/rsignell-usgs/a5c969ca1c23ab22ec34

Do you mean that `lxml` is a requirement for `OWSLib`?  
I do not see it as a requirement here:
https://github.com/geopython/OWSLib/blob/master/requirements.txt
but if it *is* a requirement, our conda-recipe isn't currently including it:
https://github.com/ioos/conda-recipes/blob/master/owslib/meta.yaml

**Comment by dblodgett-usgs** [comment]
Yeah, OWSLib does include lxml. Not sure why they put it in a REQUIREMENTS.txt, I guess because it's not always as easy as a pip install... https://github.com/geopython/OWSLib/blob/master/DEPENDENCIES.txt

**Comment by dblodgett-usgs** [comment]
p.s. Snide remarks are always welcome... :pensive:

**Comment by rsignell-usgs** [comment]
So just to make sure I understand:  If I add `lxml` to the `build` requirements in our `OWSLib` conda recipe:https://github.com/ioos/conda-recipes/blob/master/owslib/meta.yaml you think my notebook using `pyGDP` will work.   Is that correct?

@tomkralidis and @ocefpaf, just an FYI here of this issue ...

**Comment by tomkralidis** [comment]
thanks @rsignell-usgs 

FYI OWSLib does not require lxml and works with Python's default etree if no lxml is [detected](https://github.com/geopython/OWSLib/blob/master/owslib/etree.py#L35).

Does pyGDP require specify lxml functionality?

**Comment by rsignell-usgs** [comment]
@tomkralidis , I'm guessing that pyGDP must require something special, because my pyGDP test case isn't working for me with our IOOS channel OWSLib, but it works for @dblodgett-usgs (not sure how he built OWSLib)

**Comment by ocefpaf** [comment]
> I'm guessing that pyGDP must require something special

`pyGDP` requires `lxml` (https://github.com/USGS-CIDA/pyGDP/blob/master/requirements.txt#L2) and we do install it in the IOOS conda packe (https://github.com/ocefpaf/conda-recipes/blob/master/pygdp/meta.yaml#L21)

I am not sure the issue is due to the lack of `lxml`.

**Comment by dblodgett-usgs** [comment]
The function that's failing imports owslib.etree assuming that lxml will be available from it. I guess my assumption that owslib requires lxml isn't valid anymore... or never was? I don't know. 

The error is: 
---> 30     name_spaces = gml_root.nsmap
     31 
     32     attributes = []

AttributeError: 'Element' object has no attribute 'nsmap'

The offending code is here: https://github.com/USGS-CIDA/pyGDP/blob/master/pygdp/shapefile_value_handle.py#L30 

I'm afraid I'm not sure what the right fix is here... Somehow, when I import owslib.etree in my pip installed environment, lxml is available and things work in this function.

**Comment by dblodgett-usgs** [comment]
I've had someone report that downgrading to owslib 0.8.13 fixes a number of issues. @rsignell-usgs, I think you will continue to run into issues related to etree being based on base xml instead of lxml unless you fix that, and I'm not confident that just that is going to fix everything here.

**Comment by rsignell-usgs** [comment]
We rebuilt OSWlib 0.9.0 with the lxml dependency, but it did not fix the problem. 

I'll try OWSlib 0.8.13 next

**Comment by rsignell-usgs** [comment]
Okay, downgrading to owslib 0.8.13 got me further.  Now I get this error:
```
In [19]:

# Let's see what gridded datasets are available for the GDP to operate on
dataSets = myGDP.getDataSetURI()
print "Available gridded datasets:"
for d in dataSets:
    print d[0]
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-19-c87d7dea0b09> in <module>()
      1 # Let's see what gridded datasets are available for the GDP to operate on
----> 2 dataSets = myGDP.getDataSetURI()
      3 print "Available gridded datasets:"
      4 for d in dataSets:
      5     print d[0]

c:\programs\anaconda-64\envs\pygdp\lib\site-packages\pyGDP.pyc in getDataSetURI(self, anyText, CSWURL, BBox)
    161 
    162     def getDataSetURI(self, anyText='',CSWURL=CSWURL,BBox=None):
--> 163         return  webdata_handle.getDataSetURI(anyText, CSWURL, BBox)
    164 
    165     def getTimeRange(self, dataSetURI, varID, verbose=False):

c:\programs\anaconda-64\envs\pygdp\lib\site-packages\pygdp\webdata_handle.pyc in getDataSetURI(anyText, CSWURL, BBox)
     18                             csw = CatalogueServiceWeb(CSWURL, skip_caps=True)
     19                             #Works with owslib version 0.8.6.
---> 20                             csw.getrecords(keywords=[anyText], outputschema='http://www.isotc211.org/2005/gmd', esn='full', maxrecords=100)
     21                             dataSetURIs = [['title','abstract',['urls']]]
     22                             for rec in csw.records:

c:\programs\anaconda-64\envs\pygdp\lib\site-packages\owslib\csw.py in getrecords(self, qtype, keywords, typenames, propertyname, bbox, esn, sortby, outputschema, format, startposition, maxrecords, cql, xml, resulttype)
    233             self.request = node0
    234 
--> 235         self._invoke()
    236 
    237         if self.exceptionreport is None:

c:\programs\anaconda-64\envs\pygdp\lib\site-packages\owslib\csw.py in _invoke(self)
    635             self.request = util.element_to_string(self.request, encoding='utf-8')
    636 
--> 637             self.response = util.http_post(xml_post_url, self.request, self.lang, self.timeout, self.username, self.password)
    638 
    639         # parse result see if it's XML

c:\programs\anaconda-64\envs\pygdp\lib\site-packages\owslib\util.py in http_post(url, request, lang, timeout, username, password)
    371             r.add_header('Authorization', 'Basic %s' % base64string)
    372         try:
--> 373             up = urllib2.urlopen(r,timeout=timeout);
    374         except TypeError:
    375             import socket

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    152     else:
    153         opener = _opener
--> 154     return opener.open(url, data, timeout)
    155 
    156 def install_opener(opener):

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in open(self, fullurl, data, timeout)
    435         for processor in self.process_response.get(protocol, []):
    436             meth = getattr(processor, meth_name)
--> 437             response = meth(req, response)
    438 
    439         return response

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in http_response(self, request, response)
    548         if not (200 <= code < 300):
    549             response = self.parent.error(
--> 550                 'http', request, response, code, msg, hdrs)
    551 
    552         return response

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in error(self, proto, *args)
    473         if http_err:
    474             args = (dict, 'default', 'http_error_default') + orig_args
--> 475             return self._call_chain(*args)
    476 
    477 # XXX probably also want an abstract factory that knows when it makes

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in _call_chain(self, chain, kind, meth_name, *args)
    407             func = getattr(handler, meth_name)
    408 
--> 409             result = func(*args)
    410             if result is not None:
    411                 return result

c:\programs\anaconda-64\envs\pygdp\lib\urllib2.pyc in http_error_default(self, req, fp, code, msg, hdrs)
    556 class HTTPDefaultErrorHandler(BaseHandler):
    557     def http_error_default(self, req, fp, code, msg, hdrs):
--> 558         raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
    559 
    560 class HTTPRedirectHandler(BaseHandler):

HTTPError: HTTP Error 503: Service Temporarily Unavailable
```

**Comment by ocefpaf** [comment]
I found the error on`OWSLib` side and I am trying to fix it.  OWSLib fails in `from lxml import etree, ParseError`; It cannot import `ParseError`.

**Comment by ocefpaf** [comment]
BTW the error was [already fixed](https://github.com/geopython/OWSLib/commit/6f40078e0624079e266e00181b09e76c7956e733).  We need a new release of OWSLib @tomkralidis.

**Comment by paknorton** [comment]
In owslib 0.9.0 I found changing the following import in etree.py allowed lxml to load correctly but still results in other errors.

change:
   from lxml import etree, ParseError

to:
  from lxml import etree
  from lxml.etree import ParseError

**Comment by dblodgett-usgs** [comment]
@rsignell-usgs What you are doing works in our integration tests... To confirm, that is not a temporary issue with the request to ScienceBase is it? Can you use OWSLib 0.8.10 so we have comparable versions?

**Comment by paknorton** [comment]
I found downgrading to owslib 0.8.13 with:
   conda install owslib=0.8.13
fixed the errors I've been getting.

**Comment by rsignell-usgs** [comment]
There isn't a 0.8.10 built for conda, but since 0.8.13 gets us past that problem (to another but seeminly unrelated problem), I think I can stay at 0.8.13, right?

**Comment by rsignell-usgs** [comment]
I think @ayan-usgs is going to help me out here and try to get to the bottom of this.

**Comment by ocefpaf** [comment]
@rsignell-usgs we can do a `0.8.10` build if that is known to work for this specific case.  However, the error you are getting seems to be a [503](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_Server_Error).

If that is a bad error report we should strive for fixing that on OWSLib.

**Comment by ayan-usgs** [comment]
@rsignell-usgs 

Doing some sleuthing now.

**Comment by ayan-usgs** [comment]
@rsignell-usgs,

My preliminary investigates show that some times that being returned have changed between OWSLib 0.8.13 and 0.9.0. Specifically the the `getroot` method returned an `lxml.etree._Element` object in version 0.8.13, while an `xml.etree.ElementTree.Element` object in version 0.9.0. The `Element` object comes from Python's standard library and doesn't have a `nsmap` method; this leads to the heartburn that you're seeing in cell 10 of your ipython notebook.

It's unclear to me whether the change to the type of object returned by `getroot` was intentional. I'd need to study the OWSLib history. Their CHANGES.txt doc appears about of date. 

Regarding the 503 error, can you post an updated ipython notebook that goes to cell 19?

**Comment by dblodgett-usgs** [comment]
@ayan-usgs to the rescue! Thanks for the help, Andrew!

**Comment by rsignell-usgs** [comment]
@tomkralidis , regarding https://github.com/USGS-CIDA/pyGDP/issues/120#issuecomment-137441501, any chance of getting a bugfix 0.9.1 release?

**Comment by ayan-usgs** [comment]
@dblodgett-usgs: It turns out that 503 isn't coming from sciencebase, but rather http://cida.usgs.gov/gdp/geonetwork/srv/en/csw. @rsignell-usgs is using version 1.3.2 of pyGDP uses the latter URL as its default `CSWURL`. The url has been changed to reference ScienceBase at https://www.sciencebase.gov/catalog/item/54dd2326e4b08de9379b2fb1/csw'in version 1.3.3 (the current master). Do you know if http://cida.usgs.gov/gdp/geonetwork/srv/en/csw was intentionally placed in a minimally operable state?

@rsignell-usgs: You can specify `CSWURL` keyword rather than using the default when calling `getDataSetURL()` or you try upgrading to pyGDP master. The dataset at ScienceBase returns an empty dataset... I assume that's not what you're expecting.

**Comment by ayan-usgs** [comment]
@rsignell-usgs 

You may want to upgrade certifi to version 2015.04.28 as well. I'm seeing SSL validation errors using the version 14.5.14.

**Comment by tomkralidis** [comment]
All:

- OWSLib 0.9.1 has been released
- the lxml error is addressed / discussed in geopython/OWSLib/issues/276

**Comment by rsignell-usgs** [comment]
@dblodgett-usgs  and @ayan-usgs , see this comment by @tomkralidis here: https://github.com/geopython/OWSLib/issues/276#issuecomment-137605434 about adding a check in pyGDP for which version of etree is being used. 

I'm really hoping we can get [my Dustbowl example notebook](https://github.com/rsignell-usgs/python-training/blob/master/web-services/Dust_Bowl_GDP-Pandas.ipynb) working again with OWSLib 0.9.1 and a released version of pyGDP so that I can send a working version to the attendees of the Python training class last week.

**Comment by tomkralidis** [comment]
FYI for good measure: https://github.com/geopython/OWSLib/issues/276#issuecomment-138547389

**Comment by ocefpaf** [comment]
And a nice convenience function:

https://github.com/geopython/OWSLib/pull/279

**Comment by dblodgett-usgs** [comment]
@rsignell-usgs, we'll get to this sometime soon and let you know.

**Comment by rsignell-usgs** [comment]
@dblodgett-usgs , awesome.    And thanks for adding a shapefile upload to the tests.

**Comment by ayan-usgs** [comment]
@rsignell-usgs 

There's ostensibly a bug in OWSLib version 0.9.1 that took awhile to track down. See https://github.com/geopython/OWSLib/issues/280. There were a few other issues in pyGDP, but the bug caused the bulk of the heartburn.

**Comment by dblodgett-usgs** [comment]
This is being addressed in #141 

Thanks for keeping this alive @rsignell-usgs
